wallet/src/states.rs

```diff
@@ -64,319 +53,162 @@ impl<K, S> State<K, S> {
 
 impl<K, S> States<K, S>
 where
-    K: Hash + Eq,
+    K: Hash + Eq + Clone,
 {
     /// create a new States with the given initial state
     ///
     /// by default this state is always assumed confirmed
     pub fn new(key: K, state: S) -> Self {
-        let mut state = Box::new(State::new(key, state, Status::Confirmed));
-        let key_ref = KeyRef(&state.key);
-        let head: *mut State<K, S> = &mut *state;
-        let tail: *mut State<K, S> = &mut *state;
-
-        let mut map = HashMap::with_capacity(12);
-        map.insert(key_ref, state);
+        let state = StateRef::new(key.clone(), state, Status::Confirmed);
+        let mut states = LinkedHashMap::new();
+        states.insert(key, state);
 
-        Self { map, head, tail }
+        Self { states }
     }
 
     /// check wether the given state associate to this key is present
     /// in the States
     pub fn contains<Q: ?Sized>(&self, key: &Q) -> bool
     where
-        KeyRef<K>: Borrow<Q>,
+        K: Borrow<Q>,
         Q: Hash + Eq,
     {
-        self.map.contains_key(key)
+        self.states.contains_key(key)
     }
 
     /// get the underlying State associated to the given key
     pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<(&S, Status)>
     where
-        KeyRef<K>: Borrow<Q>,
+        K: Borrow<Q>,
         Q: Hash + Eq,
     {
-        self.map.get(key).map(|s| (&s.state, s.status))
+        self.states.get(key).map(|s| (&s.state, s.status))
     }
 
     /// push a new **unconfirmed** state in the States
     pub fn push(&mut self, key: K, state: S) {
-        let mut state = Box::new(State::new(key, state, Status::Pending));
-        let key_ref = KeyRef(&state.key);
-
-        state.prev = self.tail;
-        unsafe { (*self.tail).next = state.as_mut() };
-        self.tail = state.as_mut();
+        let state = StateRef::new(key.clone(), state, Status::Pending);
 
-        assert!(self.map.insert(key_ref, state).is_none());
+        assert!(self.states.insert(key, state).is_none());
     }
 
     pub fn confirm<Q: ?Sized>(&mut self, key: &Q)
     where
-        KeyRef<K>: Borrow<Q>,
+        K: Borrow<Q>,
         Q: Hash + Eq,
     {
-        if let Some(state) = self.map.get_mut(key) {
-            let state = &mut (*state) as &mut State<K, S>;
+        if let Some(state) = self.states.get_mut(key) {
             state.confirm();
         }
 
-        while self.pop_legacy_confirmed() {}
+        self.pop_old_confirmed_states()
     }
 
-    fn pop_legacy_confirmed(&mut self) -> bool {
-        let current = unsafe { &mut (*self.head) as &mut State<K, S> };
-        debug_assert!(current.confirmed());
+    fn pop_old_confirmed_states(&mut self) {
+        // maybe popping and re-inserting is better?
+        let remove: Vec<K> = self
+            .states
+            .iter()
+            .take_while(|(_, v)| v.is_confirmed())
+            .map(|(k, _)| k.clone())
+            .collect();
```

Comment:
Maybe that comment is right.

A non-allocating implementation could mark the last key in the leading contiguous sequence of confirmed entries, and then remove entries from the front until that key is met.

Reply:
re-wrote it by means of `pop_front`, with an extra insert at the end.

although now I'm thinking that the second option would probably be a bit better for the most usual case... Having to pop many states is probably not that common, it can only happen with utxo transactions, and probably even then is not the most common case. I doubt it would make a big difference anywhere, though.